File: URIComparison/py2/uri.py
# URI Comparison solutionin Python 2 for CodeEval.com by Steven A. Dunn

def get_encoded_symbol(encoded_char):
  val = int(encoded_char, 16)
  result = ""
  if val == 126: result = "~"
  return result

def parse_uri(uri):
  port_colon_pos = uri[:(uri.find(':') + 1)].find(':')
  if port_colon_pos < 0: 
    com_pos = uri.find("com")
    uri = uri[0:com_pos+3] + ":80" + uri[com_pos+4:]
  elif uri[port_colon_pos + 6] == '/':
    uri = uri[0:port_colon_pos+6] + "80" + uri[port_colon_pos+7:]

  for i in range(0, len(uri)):
    if uri[i] == '%':
      encoded_char = uri[i+1:i+3]
      uri = uri[0:i] + get_encoded_symbol(encoded_char) + uri[i+1:]
  uri = uri.lower()
  return uri

def uri_compare(uri1, uri2):
  parsed_uri1 = parse_uri(uri1)
  parsed_uri2 = parse_uri(uri2)

  print "parse uris: ", uri1, uri2

  if parsed_uri1 == parsed_uri2: return True
  else: return False

import sys

for line in open(sys.argv[1], 'r'):
  line = line.rstrip('\n').split(';')
  uri1, uri2 = line[0], line[1]

  if uri_compare(uri1, uri2): print True
  else: print False

File: URIComparison/cpp/uri.cpp
// URI Comparison solution in C++ for CodeEval.com by Steven A. Dunn

#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>

using std::cout;
using std::endl;
using std::hex;
using std::ifstream;
using std::stringstream;
using std::string;
using std::transform;

bool uriCompare(string, string);
string parseUri(string);
string getEncodedSymbol(string);

int main (int argc, char* argv[])
{
  ifstream inputFile(argv[1]);
  string line;

  if (inputFile)
  {
    while (getline(inputFile, line))
    {
      int semicolonPos = line.find(";");
      string uri1 = line.substr(0, semicolonPos);
      string uri2 = line.substr(semicolonPos + 1);
      if (uriCompare(uri1, uri2))
        cout << "True" << endl;
      else
        cout << "False" << endl;
    }
    inputFile.close();
  }

  return 0;
}

bool uriCompare(string uri1, string uri2)
{
  string parsedUri1 = parseUri(uri1);
  string parsedUri2 = parseUri(uri2);
 
  if (parsedUri1 == parsedUri2)
    return true;
  return false;
}

string parseUri(string uri)
{
  int portColonPos = uri.substr(uri.find(":") + 1).find(":");
  if (portColonPos < 0)
    uri.insert(uri.find("com") + 3, ":80");
  else if (uri.at(portColonPos + 6) == '/')
    uri.insert(portColonPos + 6, "80");

  for (int i = 0; i < uri.size(); ++i)
  {
    if (uri.at(i) == '%')
    {
      string encodedChar = uri.substr(i + 1, 2);
      uri.replace(i, 3, getEncodedSymbol(encodedChar));
    }
  }
  transform(uri.begin(), uri.end(), uri.begin(), ::tolower);  
  return uri;
}

string getEncodedSymbol(string encodedChar)
{
  stringstream ss(encodedChar);
  int val;
  ss >> hex >> val;

  string result = "";
  switch (val)
  {
    case 126: result = "~"; break;
  }

  return result;
}
